openedx/openedx-platform
# Make comprehensive theme work with django templates.

Comprehensive theming did not work with django templates (used by course wiki).

The reason it didn't work was that in order for the theme to work, theme template folder has to be added to django template dirs setting _before_ django startup.
After django startup, modifying `settings.DEFAULT_TEMPLATE_ENGINE['DIRS']` has no effect, because at that point the template engine is already initialized with a copy of the template dirs list.

Instead of running the theme startup code as an autostartup hook, we manually run it _before_ `django.setup()`. This is fine because theme startup code doesn't have to do anything else besides modifying some settings and doesn't actually need django to be initialized.

**Steps to verify**:
1. Set `COMPREHENSIVE_THEME_DIR` setting to `'/edx/app/edxapp/edx-platform/themes/red-theme/'` in your devstack (easiest way to do this is to set it in `lms.env.json`).
2. Restart LMS.
3. You will see the default logo in color and the footer will be red. Compare the content of the header and the footer on the wiki with content on the other pages (for example "Course Info").
4. Note that the wiki does not use overridden header.html and footer.html templates from the themes/red-theme/lms/templates folder, although custom logo and and custom CSS which paints the footer red both work correctly.
5. Now checkout this branch & restart LMS.
6. Clear the mako template cache by running `rm -r /tmp/mako_lms`
7. Visit the wiki page.
8. The wiki should now use overridden header.html and footer.html, just like the rest of the LMS.

Some more info with before screenshots is available at https://openedx.atlassian.net/browse/CRI-53

**Prior discussion**: https://openedx.atlassian.net/browse/OPEN-982, https://openedx.atlassian.net/browse/CRI-53
**Partner information**: 3rd party-hosted open edX instance
**JIRA ticket**: https://openedx.atlassian.net/browse/OSPR-1085
**Sandbox URLs**: TBD

---

**Settings**

``` yaml
COMPREHENSIVE_THEME_DIR: '/edx/app/edxapp/edx-platform/themes/red-theme/'
```
